### ISSUE TITLE
Support native searcher.

### ISSUE THREAD

**Issue by jcs090218** [jacktasia/dumb-jump#367]
Hi,

I wonder if this package willing to support a native searcher using [searcher](https://github.com/jcs-elpa/searcher) package? This is the package I made for native searcher solution. I think is great to have this plugin as a default searcher so it will work without any of the 3rd party searchers installed, `ag`, `rg`, `greap`, etc.

Let me know what you think! Thanks! 😄

**Comment by jacktasia** [comment]
Thanks for opening this. Your searcher package looks very interesting and I like the idea of making it a default/option, but I need to play around with it first. I’ll report back when I have. Thanks again

**Comment by phikal** [comment]
@jcs090218 I've been reading the code (especially `searcher-search-in-file`), but couldn't recognize any immediate benefits over doing something like

```elisp
(defun search-in-file (filename regexp)
  (with-temp-buffer
    (insert-file-contents-literally filename)
    (let (matches)
      (while (search-forward-regexp regexp nil t)
	(push (list :file filename
		    :string (match-string-no-properties 0)
		    :line-number (line-number-at-pos)
		    :column (current-column))
	      matches))
      matches)))
```

And from running these benchmarks ([`sqlite3.c`](https://www.sqlite.org/amalgamation.html) is 7.8MB large):

```elisp
(benchmark-run 1
  (search-in-file "~/sqlite3.c" "int"))
;;; (215.002425134 1 0.03455014099999998)

(benchmark-run 1
  (searcher-search-in-file "~/sqlite3.c" "int"))
;; (214.335509272 2 0.06406018300000005)
```

The performance doesn't look that great (for either). My guess is that the repeated invocation of `line-number-at-pos` makes the runtime polynomial, even with simple regular expressions (if you drop the function call, the performance increases significantly). Do you have any plans to increase the performance?

**Comment by jcs090218** [comment]
> The performance doesn't look that great (for either). My guess is that the repeated invocation of line-number-at-pos makes the runtime polynomial, even with simple regular expressions (if you drop the function call, the performance increases significantly). Do you have any plans to increase the performance?

Yeah, I do have the plan for increasing the performance and all that. Yet, currently I don't have a solution yet. By the way, thanks for the test! I will definitely look into it and see what I can do to the large file performance.

**Comment by phikal** [comment]
That's great to see. But from my tests, it seems 10-20% of the CPU time is still "wasted" on garbage collection. Is there are reason you convert the buffer into a string, instead of directly using the buffer? I can imagine that allocating at least twice the necessary memory could burden the garbage collector more than necessary.

---

Here the updated test results:

```elisp
(benchmark-run 100
  (search-in-file "~/sqlite3.c" "SQLITE_[[:word:]_]+?AUTO[[:word:]]+"))
;; (8.661332335 1 0.04296737199999967)

(benchmark-run 100
  (searcher-search-in-file "~/sqlite3.c" "SQLITE_[[:word:]_]+?AUTO[[:word:]]+"))
;; (11.833157401 50 1.9641602009999994)
``

Notice that `searcher-search-in-file` invokes GC 50 times, so on average every two searches, while the buffer-based approach only has to do so once per 100 invocations.

**Comment by jcs090218** [comment]
> Is there are reason you convert the buffer into a string, instead of directly using the buffer?

Good point here. I actually have managed to use buffer yet it actually much slow on my machine! I have tested yesterday night and today twice but no luck improving the performance with using the buffer directly. Any idea?

> But from my tests, it seems 10-20% of the CPU time is still "wasted" on garbage collection.

Since the experiment with buffer wasn't going well; I have plan to just raise `gc-cons-threshold`. But this also failed by wrapping it inside the package itself. See [here](https://github.com/jcs-elpa/searcher/blob/493cfcd4bdcac8e64666552622d5eafee3138fb5/searcher.el) for implementation.

But with this, it works.

```el
(let ((gc-cons-threshold 400000000))
  (message
   "%s"
   (benchmark-run test-performance--loop
     (searcher-search-in-file test-performance--file "int"))))
```

I was attempting to resolve it within my package but I realized once the `gc-cons-threshold` pop off from the stack then the GC triggers immediately, hence the output make no differences.

**Comment by jacktasia** [comment]
Thanks for the suggestion @jcs090218 and the great benchmarking discussion @phikal.                                                                                 

 Going to close this out since a pure-elisp searcher is just too slow for an interactive jump-to-definition tool, even after the v0.2.0 improvements. The GC pressure is also a structural concern. And since `grep` already serves as a universal zero-install fallback (ag → rg → gnu-grep → grep), there isn't really a gap to fill here. Appreciate the idea though!
